drivers/can/stm32_can.c

```diff
@@ -0,0 +1,765 @@
+		scale_masked &&  mode_masked ? CAN_FILTER_EXTENDED :
+		CAN_FILTER_EXTENDED_MASKED;
+
+	return type;
+}
+
+static void can_stm32_set_filter_bank(int filter_nr,
+									  CAN_FilterRegister_TypeDef *filter_reg,
+									  enum can_filter_type filter_type,
+									  u32_t id, u32_t mask)
+{
+	switch (filter_type) {
+	case CAN_FILTER_STANDARD:
+		switch (filter_nr & 0x03) {
+		case 0:
+			filter_reg->FR1 = (filter_reg->FR1 & 0xFFFF0000) | id;
+			break;
+		case 1:
+			filter_reg->FR1 = (filter_reg->FR1 & 0x0000FFFF)
+					  | (id << 16);
+			break;
+		case 2:
+			filter_reg->FR2 = (filter_reg->FR2 & 0xFFFF0000) | id;
+			break;
+		case 3:
+			filter_reg->FR2 = (filter_reg->FR2 & 0x0000FFFF)
+					  | (id << 16);
+			break;
+		}
+		break;
+	case CAN_FILTER_STANDARD_MASKED:
+		switch (filter_nr & 0x02) {
+		case 0:
+			filter_reg->FR1 = id | (mask << 16);
+			break;
+		case 2:
+			filter_reg->FR2 = id | (mask << 16);
+			break;
+		}
+		break;
+	case CAN_FILTER_EXTENDED:
+		switch (filter_nr & 0x02) {
+		case 0:
+			filter_reg->FR1 = id;
+			break;
+		case 2:
+			filter_reg->FR2 = id;
+			break;
+		}
+		break;
+	case CAN_FILTER_EXTENDED_MASKED:
+		filter_reg->FR1 = id;
+		filter_reg->FR2 = mask;
+		break;
+	}
+}
+
+static inline int can_stm32_set_filter(struct can_filter *filter,
+				       struct can_stm32_data *device_data,
+				       CAN_TypeDef *can,
+				       int *filter_index)
+{
+	u32_t mask = 0;
+	u32_t id;
+	int filter_nr = 0;
+	int filter_index_tmp = CAN_NO_FREE_FILTER;
+	int bank_nr;
+	u32_t bank_bit;
+	enum can_filter_type filter_type;
+
+	if (filter->id_type == CAN_STANDARD_IDENTIFIER) {
+		id = (filter->std_id << CAN_FIRX_STD_ID_POS)
+			 | (filter->rtr << CAN_FIRX_STD_RTR_POS);
+
+		if (filter->std_id_mask == CAN_STD_ID_MASK && filter->rtr_mask) {
+			filter_type = CAN_FILTER_STANDARD;
+		} else {
+			filter_type = CAN_FILTER_STANDARD_MASKED;
+			mask = (filter->std_id_mask << CAN_FIRX_STD_ID_POS)
+			       | (filter->rtr_mask << CAN_FIRX_STD_RTR_POS)
+			       | (1U << CAN_FIRX_STD_IDE_POS);
+		}
+	} else {
+		id = (filter->ext_id << CAN_FIRX_EXT_EXT_ID_POS)
+		     | (filter->rtr << CAN_FIRX_EXT_RTR_POS)
+		     | (1U << CAN_FIRX_EXT_IDE_POS);
+
+		if (filter->ext_id_mask == CAN_EXT_ID_MASK && filter->rtr_mask) {
+			filter_type = CAN_FILTER_EXTENDED;
+		} else {
+			filter_type = CAN_FILTER_EXTENDED_MASKED;
+			mask = (filter->ext_id_mask << CAN_FIRX_EXT_EXT_ID_POS)
+			       | (filter->rtr_mask << CAN_FIRX_EXT_RTR_POS)
+			       | (1U << CAN_FIRX_EXT_IDE_POS);
+		}
+	}
+	SYS_LOG_DBG("Setting filter ID: 0x%x, mask: 0x%x", filter->ext_id,
+		    filter->ext_id_mask);
+	SYS_LOG_DBG("Filter type: %s ID %s mask (%d)",
+		    (filter_type == CAN_FILTER_STANDARD ||
+		     filter_type == CAN_FILTER_STANDARD_MASKED) ?
+		    "standard" : "extended",
+		    (filter_type == CAN_FILTER_STANDARD_MASKED ||
+		     filter_type == CAN_FILTER_EXTENDED_MASKED) ?
+		    "with" : "without",
+		    filter_type);
+	do {
```

Comment:
hum, that's an ugly code block with too many if/if/if...
Could you rewrite this? You can externalize some logic to inline functions if you want.

 (btw, why not using a switch on filter_type here? As you are doing at line 581)

Reply:
I will try to reduce it to 2 levels.
The switch is not used because of the break.

Reply:
Done. I managed to get rid of the ifs acting like a switch.